### PR TITLE
Improvements related to LeftShift/Reduce cancelation.

### PR DESCRIPTION
**Pull request by sellout**
- handle ZipMapKeys when canceling;
- simplify `Reduce(LeftShift)` pairs;
- restrict `MapFunc.foldConstant` to guarantee that it can only return a
  constant;
- rename `MapFunc.normalize` to `rewrite`, and create a new `normalize`
  that composes `rewrite` with `foldConstant`;
- upgrade `ReduceFunc` from `Traverse` to `Traverse1`; and
- remove some duplication – including adding `::\::` and `::/::` for
  Injectable instances.

### CODE REVIEW

File: connector/src/test/scala/quasar/qscript/ShiftReadSpec.scala [hunk 1]
```diff
@@ -38,9 +43,25 @@ class ShiftReadSpec extends quasar.Qspec with QScriptHelpers {
...
-      Fix(QS.inject(QC.inj(Map(
-        Fix(SR.inj(Const[ShiftedRead, Fix[QST]](ShiftedRead(sampleFile, IncludeId)))),
-        Free.roll(ProjectIndex(HoleF, IntLit(1)))))))
+      Fix(QCT.inj(Map(
+        Fix(SRT.inj(Const[ShiftedRead, Fix[QST]](ShiftedRead(sampleFile, IncludeId)))),
+        Free.roll(ProjectIndex(HoleF, IntLit(1))))))
     }
   }
+
+  "shift a simple aggregated read" in {
```
Comment (sellout) [inline connector/src/test/scala/quasar/qscript/ShiftReadSpec.scala]: Here’s your example query.

**Comment by sellout** [comment]
Fixes #1480.

**Comment by wemrysi** [comment]
@sellout Any idea why the simplification doesn't appear to be applied to the branches of a `ThetaJoin`?

QUERY
`select pop from /explain/data`

LP
```
 Optimized
 Invoke(SQUASH)
 ╰─ Invoke(MAKE_OBJECT)
    ├─ Str("pop")
    ╰─ ObjectProject(/explain/data["pop"])
 
 Typechecked
 Invoke(SQUASH)
 ╰─ Invoke(MAKE_OBJECT)
    ├─ Str("pop")
    ╰─ Let('__check0)
       ├─ Read(/explain/data)
       ╰─ Typecheck(Type(Obj(Map(),Some(Top))))
          ├─ Free('__check0)
          ├─ ObjectProject('__check0["pop"])
          ╰─ Constant(NA)
```

QSCRIPT
```
QScript
 List
 ├─ Unreferenced
 ╰─ ThetaJoin((),LeftShift(Read(Read(/quasar-test/run_afda694f/explain/data)),ZipMapKeys(SrcHole),ConcatArrays(MakeArray(LeftSide), MakeArray(RightSide))),LeftShift(Read(Read(/quasar-test/run_afda694f/explain/data)),SrcHole,ProjectField(RightSide, Constant(Str(pop)))),Constant(Bool(true)),Inner,MakeMap(Constant(Str(pop)), Guard(ProjectIndex(ProjectIndex(LeftSide, Constant(Int(1))), Constant(Int(1))), Type(Obj(Map(),Some(Top))), RightSide, Undefined())))
    ╰─ ExternallyManaged(Extern)
 
 QScript (ShiftRead)
 List
 ├─ Unreferenced
 ╰─ ThetaJoin((),LeftShift(Reduce(ShiftedRead(ShiftedRead(/quasar-test/run_afda694f/explain/data, IdStatus(IncludeId))),Constant(Null()),[UnshiftMap(ProjectIndex(SrcHole, Constant(Int(0))), ProjectIndex(SrcHole, Constant(Int(1))))],ReduceIndex(0)),ZipMapKeys(SrcHole),ConcatArrays(MakeArray(LeftSide), MakeArray(RightSide))),Map(ShiftedRead(ShiftedRead(/quasar-test/run_afda694f/explain/data, IdStatus(IncludeId))),ProjectField(ProjectIndex(SrcHole, Constant(Int(1))), Constant(Str(pop)))),Constant(Bool(true)),Inner,MakeMap(Constant(Str(pop)), Guard(ProjectIndex(ProjectIndex(LeftSide, Constant(Int(1))), Constant(Int(1))), Type(Obj(Map(),Some(Top))), RightSide, Undefined())))
    ╰─ ExternallyManaged(Extern)
```

**Comment by sellout** [comment]
Oh, damn – no, that should be working – it’s called in `applyAll` and `applyFreeQS`, which should cover all the cases. I’ll dig into this one.

**Comment by sellout** [comment]
@wemrysi Just to clarify – you checked out this branch and ran that query against it? Because that query certainly wouldn’t have canceled without this PR.

**Comment by wemrysi** [comment]
@sellout Yes, I pulled it locally into one of my branches to see how it improved things.

**Comment by sellout** [comment]
Woah, wait … it applied it to the _right_ branch, but not the _left_ one. That makes this look like an easy bug …

**Comment by wemrysi** [comment]
Oh, indeed, good catch!

**Comment by sellout** [comment]
Ah, it looks like the left side there is more complicated – basically, it can’t be canceled out in its current form because it references the `LeftSide` of the LeftShift in its repair function. But, I can see how another normalization (that would recognize that the `LeftSide` is never used and thus eliminate it) would allow it to be applied. It’s not a bug in this particular transformation, though.

**Comment by sellout** [comment]
I’m about to make a change that _should_ prevent that ThetaJoin from ever being created, though *crosses fingers*.

**Comment by wemrysi** [comment]
@sellout LGTM modulo conflicts.

**Pull request merged**